src/compiler/checker.ts

```diff
@@ -10594,6 +10599,15 @@ module ts {
                         checkIndexConstraints(type);
                     }
                 }
+
+                if (symbol && symbol.declarations) {
+                    for (let declaration of symbol.declarations) {
+                        if (declaration.kind === SyntaxKind.ClassDeclaration && !(declaration.flags & NodeFlags.Ambient)) {
+                            error(node, Diagnostics.An_interface_cannot_merge_with_a_non_ambient_class);
+                            break;
+                        }
+                    }
+                }

```

Comment:
This doesn't work for 

```ts
interface I {}
declare class I {}
module I {}
```
Since the module occupies the value space, the blanket check looking at the symbol will trigger an error, but the merge should be able to go through.